src/coreclr/jit/optimizer.cpp

```diff
@@ -5918,10 +5918,33 @@ void Compiler::optRemoveRedundantZeroInits()
 
     assert(fgNodeThreading == NodeThreading::AllTrees);
 
-    for (BasicBlock* block = fgFirstBB; (block != nullptr) && !block->HasFlag(BBF_MARKED);
-         block             = block->GetUniqueSucc())
+    for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->GetUniqueSucc())
     {
-        block->SetFlags(BBF_MARKED);
+        if (m_dfsTree->HasCycle())
+        {
+            // See if this block is a cycle entry
+            //
+            bool stop = false;
+            for (BasicBlock* const pred : block->PredBlocks())
```

Comment:
This could use `BlockPredsWithEH`... It probably doesn't actually matter since we walk forwards with `block->GetUniqueSucc()`, so won't ever end up at a handler, but it would be more consistent with the graph that the DFS tree is constructed over.

Reply:
Done... no change in SPMI diffs.